projects/libhtp/fuzz.diff

```diff
@@ -0,0 +1,294 @@
+diff --git a/test/Makefile.am b/test/Makefile.am
+index fb14b94..2d57e7b 100644
+--- a/test/Makefile.am
++++ b/test/Makefile.am
+@@ -9,12 +9,15 @@ AM_CPPFLAGS = -I$(top_srcdir) -I$(top_builddir)/htp -Wno-write-strings -DGTEST_U
+ 
+ AUTOMAKE_OPTIONS = subdir-objects
+ EXTRA_DIST = files
+-check_PROGRAMS = test_all
++check_PROGRAMS = test_all test_fuzz
+ check_LIBRARIES = libgtest.a
+ 
+ test_all_SOURCES = test_bstr.cpp test_gunzip.cpp test_hybrid.cpp test_main.cpp test_multipart.cpp test.c test.h test_utils.cpp test_bench.cpp
+ test_all_LDADD = libgtest.a -lpthread $(LDADD)
+ 
++test_fuzz_SOURCES = fuzz/onefile.c fuzz/fuzz_htp.c test.c
++test_fuzz_LDADD = $(LDADD)
++
+ libgtest_a_SOURCES = gtest/gtest-all.cc gtest/gtest_main.cc gtest/gtest.h
+ 
+ TESTS_ENVIRONMENT = srcdir=$(srcdir)/files
+diff --git a/test/fuzz/fuzz_htp.c b/test/fuzz/fuzz_htp.c
+new file mode 100644
+index 0000000..b9c274a
+--- /dev/null
++++ b/test/fuzz/fuzz_htp.c
+@@ -0,0 +1,157 @@
++/**
++ * @file
++ * @author Philippe Antoine <contact@catenacyber.fr>
++ * fuzz harness for libhtp
++ */
++
++
++#include <errno.h>
++#include <stdlib.h>
++#include <sys/types.h>
++#include <string.h>
++#include <stdio.h>
++
++#include <sys/stat.h>
++#include <fcntl.h>
++
++#include "htp/htp.h"
++#include "test/test.h"
++
++
++FILE * logfile = NULL;
++
++
++/**
++ * Invoked at the end of every transaction. 
++ *
++ * @param[in] connp
++ */
++static int callback_response(htp_tx_t *out_tx) {
++    if (out_tx != NULL) {
++        char *x = bstr_util_strdup_to_c(out_tx->request_line);
++        fprintf(logfile, "%s\n", x);
++        free(x);
++    }
++    return 0;
++}
++
++/**
++ * Invoked every time LibHTP wants to log. 
++ *
++ * @param[in] log
++ */
++static int callback_log(htp_log_t *log) {
++    fprintf(logfile, "[%d][code %d][file %s][line %d] %s\n",
++        log->level, log->code, log->file, log->line, log->msg);
++    return 0;
++}
++
++void fuzz_openFile(const char * name) {
++    if (logfile != NULL) {
++        fclose(logfile);
++    }
++    logfile = fopen(name, "w");
++}
++
++int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
++    htp_cfg_t *cfg;
++    htp_connp_t * connp;
++    int rc;
++    test_t test;
++
++    //initialize output file
++    if (logfile == NULL) {
++        logfile = fopen("/dev/null", "w");
++        if (logfile == NULL) {
++            return 0;
```

Comment:
Should probably `abort()` here since fuzzing can't happen if we can't get a file handle to `/dev/null` right?

Reply:
I guess this would be a more appropriate behavior indeed, I will do another pull request there